core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,69 +1354,371 @@ class ReassignedPartitionsIsrChangeListener(protected val controller: KafkaContr
+    }
+  }
+
+  case object UpdateMetrics extends ControllerEvent {
+    override def process(): Unit = {
+      val opc = if (!isActive)
+        0
+      else
+        controllerContext.partitionLeadershipInfo.count(p =>
+          !controllerContext.liveOrShuttingDownBrokerIds.contains(p._2.leaderAndIsr.leader) &&
+            !topicDeletionManager.isTopicQueuedUpForDeletion(p._1.topic)
+        )
+      offlinePartitionCount.set(opc)
+
+      val pric = if (!isActive)
+        0
+      else
+        controllerContext.partitionReplicaAssignment.count { case (topicPartition, replicas) =>
+          controllerContext.partitionLeadershipInfo(topicPartition).leaderAndIsr.leader != replicas.head &&
+            !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic)
+        }
+      preferredReplicaImbalanceCount.set(pric)
+    }
+  }
+
+  private def triggerControllerMove(): Unit = {
+    activeControllerId.set(-1)
+    controllerContext.zkUtils.deletePath(ZkUtils.ControllerPath)
+  }
+}
+
+/**
+  * This is the zookeeper listener that triggers all the state transitions for a replica
+  */
+class BrokerChangeListener(protected val controller: KafkaController, controllerEventQueue: LinkedBlockingQueue[ControllerEvent]) extends IZkChildListener with Logging {
+  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
+    import scala.collection.JavaConverters._
+    controllerEventQueue.put(controller.BrokerChange(currentChilds.asScala))
+  }
+}
+
+class TopicChangeListener(protected val controller: KafkaController, controllerEventQueue: LinkedBlockingQueue[ControllerEvent]) extends IZkChildListener with Logging {
+  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
+    import scala.collection.JavaConverters._
+    controllerEventQueue.put(controller.TopicChange(currentChilds.asScala.toSet))
+  }
+}
+
+class PartitionModificationsListener(protected val controller: KafkaController, controllerEventQueue: LinkedBlockingQueue[ControllerEvent], topic: String) extends IZkDataListener with Logging {
+  override def handleDataChange(dataPath: String, data: scala.Any): Unit = {
+    controllerEventQueue.put(controller.PartitionModifications(topic))
+  }
+
+  override def handleDataDeleted(dataPath: String): Unit = {}
+}
+
+/**
+  * Delete topics includes the following operations -
+  * 1. Add the topic to be deleted to the delete topics cache, only if the topic exists
+  * 2. If there are topics to be deleted, it signals the delete topic thread
+  */
+class TopicDeletionListener(protected val controller: KafkaController, controllerEventQueue: LinkedBlockingQueue[ControllerEvent]) extends IZkChildListener with Logging {
+  override def handleChildChange(parentPath: String, currentChilds: java.util.List[String]): Unit = {
+    import scala.collection.JavaConverters._
+    controllerEventQueue.put(controller.TopicDeletion(currentChilds.asScala.toSet))
```

Comment:
Hmm, there is a slight change of behavior here. Earlier, if a topic deletion is initiated, it will be processed immediately. Now, the topic deletion will only be processed when the scheduler adds a topic deletion event. This means that topic deletion could be delayed by up to 5 seconds, which is a degradation.